src/main/java/hu/akarnokd/rxjava2/operators/FlowableWindowPredicate.java

```diff
@@ -0,0 +1,206 @@
+
+    enum Mode {
+        /** The item triggering the new buffer will be part of the new buffer. */
+        BEFORE,
+        /** The item triggering the new buffer will be part of the old buffer. */
+        AFTER,
+        /** The item won't be part of any buffers. */
+        SPLIT
+    }
+
+    final Publisher<T> source;
+
+    final Predicate<? super T> predicate;
+
+    final Mode mode;
+
+    final int bufferSize;
+
+    FlowableWindowPredicate(Publisher<T> source, Predicate<? super T> predicate, Mode mode,
+            int bufferSize) {
+        this.source = source;
+        this.predicate = predicate;
+        this.mode = mode;
+        this.bufferSize = bufferSize;
+    }
+
+    @Override
+    protected void subscribeActual(Subscriber<? super Flowable<T>> s) {
+        source.subscribe(new WindowPredicateSubscriber<T>(s, predicate, mode, bufferSize));
+    }
+
+    @Override
+    public Publisher<Flowable<T>> apply(Flowable<T> upstream) {
+        return new FlowableWindowPredicate<T>(upstream, predicate, mode, bufferSize);
+    }
+
+    static final class WindowPredicateSubscriber<T>
+    extends AtomicInteger
+    implements ConditionalSubscriber<T>, Subscription, Runnable {
+
+        final Subscriber<? super Flowable<T>> actual;
+
+        final Predicate<? super T> predicate;
+
+        final Mode mode;
+
+        final int bufferSize;
+
+        final AtomicBoolean cancelOnce;
+
+        Subscription s;
+
+        UnicastProcessor<T> window;
+
+        WindowPredicateSubscriber(Subscriber<? super Flowable<T>> actual,
+                Predicate<? super T> predicate, Mode mode,
+                int bufferSize) {
+            super(1);
+            this.actual = actual;
+            this.predicate = predicate;
+            this.mode = mode;
+            this.bufferSize = bufferSize;
+            this.cancelOnce = new AtomicBoolean();
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            if (SubscriptionHelper.validate(this.s, s)) {
+                this.s = s;
+                actual.onSubscribe(this);
+            }
+        }
+
+        @Override
+        public void onNext(T t) {
+            if (!tryOnNext(t)) {
+                s.request(1);
+            }
+        }
+
+        @Override
+        public boolean tryOnNext(T t) {
+            UnicastProcessor<T> w = window;
+            if (w == null) {
+                w = newWindow();
+            }
+
+            boolean b;
+
+            try {
+                b = predicate.test(t);
+            } catch (Throwable ex) {
+                Exceptions.throwIfFatal(ex);
+                s.cancel();
+                actual.onError(ex);
+                return true;
+            }
+
+            if (b) {
+                if (mode == Mode.AFTER) {
+                    w.onNext(t);
+                }
+                w = newWindow();
```

Comment:
If `cancelOnce` is true, no new window should be created or emitted but the active window should be allowed to receive items.